### PR TITLE
First dbt Cloud Task

### LINKED ISSUE #6: Create the boilerplate code for the integration
get some basic api calls to dbt Cloud working

### PR DESCRIPTION
**Pull request by desertaxle**
## Summary

Introduces the first dbt Cloud task to trigger a job run: `trigger_job_run`. Sets up the necessary boilerplate for interacting with the dbt Cloud administrative API.

Currently, jobs can only be triggered via this collection, but not waited on. A future PR will introduce the `get_run` task which will be used to wait for job run completion.

## Relevant Issue(s)

Closes #6 

## Checklist
- [x] Summarized PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-dbt/blob/main/CHANGELOG.md)

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -36,6 +36,9 @@ plugins:
 
 nav:
     - Home: index.md
-    - Tasks: tasks.md
-    - Flows: flows.md
+    - Cloud:
+      - Credentials: cloud/credentials.md
```
Comment (ahuang11) [inline mkdocs.yml:40]: Does dbt core and cloud use separate credentials?

File: prefect_dbt/cloud/clients.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Sends a request to the [trigger job run endpoint](https://docs.getdbt.com/dbt-cloud/api-v2#tag/Jobs/operation/triggerRun)
+        to initiate a job run.
+
+        Args:
+            job_id: The ID of the job to trigger.
+            options: An optional TriggerJobRunOptions instance to specify overrides for the triggered job run.
+
+        Returns:
+            The response from the dbt Cloud administrative API.
+        """  # noqa
```
Comment (ahuang11) [inline prefect_dbt/cloud/clients.py:45]: I think the docs still work if you start a new line.

File: prefect_dbt/cloud/clients.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        if options is None:
+            options = TriggerJobRunOptions()
+
+        json = options.dict(exclude_none=True)
+        print(json)
```
Comment (ahuang11) [inline prefect_dbt/cloud/clients.py:50]: Remove print maybe?

File: prefect_dbt/cloud/jobs.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    Args:
+        dbt_cloud_credentials: Credentials for authenticating with dbt Cloud.
+        job_id: The ID of the job to trigger.
+        options: An optional TriggerJobRunOptions instance to specify overrides
+            for the triggered job run.
+
+    Returns:
+        The response from the dbt Cloud administrative API.
+    """
```
Comment (ahuang11) [inline prefect_dbt/cloud/jobs.py]: Add an example maybe?

**Review by ahuang11 — CHANGES_REQUESTED** [review]

File: prefect_dbt/cloud/clients.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        response = await self._admin_client.post(
+            url=f"/jobs/{job_id}/run/",
+            json=options.dict(exclude_none=True),
```
Comment (ahuang11) [inline prefect_dbt/cloud/clients.py:53]: Use json=json

Comment (desertaxle) [reply prefect_dbt/cloud/clients.py:51]: That `json` variable was just for debugging so I've removed those two lines.

File: prefect_dbt/cloud/clients.py (hunk 2 above)
Comment (desertaxle) [reply prefect_dbt/cloud/clients.py:45]: I have a long URL that all needs to on one line higher up in the docstring. That's what I need the `# noqa` comment for.

File: mkdocs.yml (hunk 1 above)
Comment (sungchun12) [reply mkdocs.yml:40]: The main different in credentials is that dbt Cloud will use API service tokens while dbt Core will use `profiles.yml` credentials.

**Comment by sungchun12** [comment]
Heads up I'll review this in full tomorrow!

File: prefect_dbt/cloud/clients.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class DbtCloudAdministrativeClient:
+    """
+    Client for interacting with the dbt cloud Administrative API.
+
+    Args:
+        api_key: API key to authenticate with the dbt Cloud administrative API.
+        account_id: ID of dbt Cloud account with which to interact.
+        domain: Domain at which the dbt Cloud API is hosted.
```
Comment (sungchun12) [inline prefect_dbt/cloud/clients.py:17]: Good callout on making this a parameter as we have single tenant customers!

File: prefect_dbt/cloud/clients.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        )
+
+        response.raise_for_status()
+
+        return response
+
+    async def __aenter__(self):
+        if self._closed:
+            raise RuntimeError(
+                "The client cannot be stared again after it has be closed."
```
Comment (sungchun12) [inline prefect_dbt/cloud/clients.py:61]:
```suggestion
                "The client cannot be started again after it has been closed."
```

File: prefect_dbt/cloud/clients.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    """
+
+    def __init__(self, api_key: str, account_id: int, domain: str):
+        self._closed = False
+        self._started = False
+
+        self._admin_client = AsyncClient(
+            headers={
+                "Authorization": f"Bearer {api_key}",
+                "user-agent": f"prefect-{prefect.__version__}",
```
Comment (sungchun12) [inline prefect_dbt/cloud/clients.py:27]: When you're ready to test dry runs with this user-agent, I should see something in the logs come through. 

Looks like the 1.0 prefect user-agent is only coming through for now.

image

**Comment by sungchun12** [comment]
After you resolve the spelling issue and default cause comments above, feel free to squash merge!

File: prefect_dbt/cloud/clients.py (hunk 8 above)
Comment (desertaxle) [reply prefect_dbt/cloud/clients.py:27]: I ran a test against the dbt Cloud API so there should be at lease one request with `prefect-2.0b5` as the user-agent header.

**Comment by desertaxle** [comment]
> After you resolve the spelling issue and default cause comments above, feel free to squash merge!

@sungchun12 I don't see a comment about the default cause. Would you be able to submit it again?

File: prefect_dbt/cloud/models.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+from pydantic import BaseModel, Field
+
+
+class TriggerJobRunOptions(BaseModel):
+    """
+    Defines options that can be defined when triggering a dbt Cloud job run.
+    """
+
+    cause: str = Field(
+        default="Kicked off from Prefect",
```
[the remark this answers is not in the document]
Comment (sungchun12) [reply prefect_dbt/cloud/models.py:13]: @desertaxle do you see the above requested change now?

**Comment by sungchun12** [comment]
FYI I noticed that you're using dbt `0.21.0` I recommend updating it to `1.1(latest)` to ensure a realistic environment where people trigger dbt Cloud jobs.

Not needed to merge this PR, but wanted to write it down for posterity.
image

File: prefect_dbt/cloud/clients.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
+"""Module containing clients for interacting with the dbt Cloud API"""
+from typing import Optional
+
+import prefect
+from httpx import AsyncClient, Response
+
+from prefect_dbt.cloud.models import TriggerJobRunOptions
+
+
+class DbtCloudAdministrativeClient:
```
Comment (ahuang11) [inline prefect_dbt/cloud/clients.py:10]: Can you explain the purpose of this client class?

Is this class's purpose only to implement async context manager?  If so, I think this works:
```
from dataclasses import dataclass
from httpx import AsyncClient

import prefect


@dataclass
class DbtCloudCredentials():
    
    api_key: str
    account_id: int
    domain: str = "cloud.getdbt.com"

    def get_client(self):
        return AsyncClient(
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "user-agent": f"prefect-{prefect.__version__}",
            },
            base_url=f"https://{self.domain}/api/v2/accounts/{self.account_id}",
        )


credentials = DbtCloudCredentials("api_key", "account_id")

async with credentials.get_client() as client:
    print(client)
```

File: prefect_dbt/cloud/models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (desertaxle) [reply prefect_dbt/cloud/models.py]: I updated the default cause to include the name of the task run that triggered the job run.

**Review by sungchun12 — DISMISSED** [review]

File: CHANGELOG.md [hunk 11]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- `trigger_job_run` task - [#1](https://github.com/PrefectHQ/prefect-dbt/pull/16)
```
Comment (ahuang11) [inline CHANGELOG.md:12]:
```suggestion
- `trigger_job_run` task - [#16](https://github.com/PrefectHQ/prefect-dbt/pull/16)
```

**Review by ahuang11 — CHANGES_REQUESTED** [review]

**Review by ahuang11 — APPROVED** [review]

**Pull request merged**